bindings/matrix-sdk-ffi/src/timeline.rs

```diff
@@ -0,0 +1,409 @@
+            _ => None,
+        }
+    }
+
+    pub fn push(self: Arc<Self>) -> Option<Arc<TimelineItem>> {
+        unwrap_or_clone_arc_into_variant!(self, .0, VecDiff::Push { value } => value)
+    }
+}
+
+pub struct InsertAtData {
+    pub index: u32,
+    pub item: Arc<TimelineItem>,
+}
+
+pub struct UpdateAtData {
+    pub index: u32,
+    pub item: Arc<TimelineItem>,
+}
+
+pub struct MoveData {
+    pub old_index: u32,
+    pub new_index: u32,
+}
+
+#[derive(Clone, Copy)]
+pub enum TimelineChange {
+    Replace,
+    InsertAt,
+    UpdateAt,
+    RemoveAt,
+    Move,
+    Push,
+    Pop,
+    Clear,
+}
+
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct TimelineItem(matrix_sdk::room::timeline::TimelineItem);
+
+impl TimelineItem {
+    fn from_arc(arc: Arc<matrix_sdk::room::timeline::TimelineItem>) -> Arc<Self> {
+        // SAFETY: This is valid because Self is a repr(transparent) wrapper
+        //         around the other Timeline type.
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as _) }
+    }
+
+    pub fn as_event(self: Arc<Self>) -> Option<Arc<EventTimelineItem>> {
+        use matrix_sdk::room::timeline::TimelineItem as Item;
+        unwrap_or_clone_arc_into_variant!(self, .0, Item::Event(evt) => {
+            Arc::new(EventTimelineItem(evt))
+        })
+    }
+
+    pub fn as_virtual(self: Arc<Self>) -> Option<Arc<VirtualTimelineItem>> {
+        use matrix_sdk::room::timeline::TimelineItem as Item;
+        unwrap_or_clone_arc_into_variant!(self, .0, Item::Virtual(vt) => {
+            Arc::new(VirtualTimelineItem(vt))
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct EventTimelineItem(matrix_sdk::room::timeline::EventTimelineItem);
+
+impl EventTimelineItem {
+    pub fn key(&self) -> TimelineKey {
+        self.0.key().into()
+    }
+
+    pub fn sender(&self) -> String {
+        self.0.sender().to_string()
+    }
+
+    pub fn content(&self) -> Arc<TimelineItemContent> {
+        Arc::new(TimelineItemContent(self.0.content().clone()))
+    }
+
+    pub fn reactions(&self) -> Vec<Reaction> {
+        self.0
+            .reactions()
+            .iter()
+            .map(|(k, v)| Reaction {
+                key: k.to_owned(),
+                details: v
+                    .iter()
+                    .map(|details| ReactionDetails {
+                        id: (&details.id).into(),
+                        sender: details.sender.to_string(),
+                    })
+                    .collect(),
+            })
+            .collect()
+    }
+
+    pub fn origin_server_ts(&self) -> u64 {
+        self.0.origin_server_ts().0.into()
+    }
+}
+
+#[derive(Clone)]
+pub struct TimelineItemContent(matrix_sdk::room::timeline::TimelineItemContent);
+
+impl TimelineItemContent {
+    pub fn as_message(self: Arc<Self>) -> Option<Arc<Message>> {
+        use matrix_sdk::room::timeline::TimelineItemContent as C;
+        unwrap_or_clone_arc_into_variant!(self, .0, C::Message(msg) => Arc::new(Message(msg)))
```

Comment:
Double checking for my own knowledge, is this and others like `as_text`/`as_image` an expensive operation if it gets called repeatedly?

Reply:
Unfortunately, I expect it to clone the entire inner message every time. Whether that fits your definition of expensive, I don't know ^^